src/ll/gcd.rs

```diff
@@ -0,0 +1,76 @@
+use ll;
+use mem;
+use ll::limb::Limb;
+
+pub unsafe fn gcd(gp: *mut Limb, ap: *mut Limb, an: i32, bp: *mut Limb, bn: i32) -> i32 {
+    let mut g = &mut Limb(1);
+
+    while is_even(ap) && is_even(bp) && !ll::is_zero(ap, an) && !ll::is_zero(bp, bn) {
+        ll::shr(ap, ap, an, 1);
+        ll::shr(bp, bp, bn, 1);
+        *g = *g << 1;
+    }
+
+    let mut tmp = mem::TmpAllocator::new();
+    let t = tmp.allocate(an as usize);
+
+    while !ll::is_zero(ap, an) {
+
+        while is_even(ap) && !ll::is_zero(ap, an) {
+            ll::shr(ap, ap, an, 1);
+        }
+
+        while is_even(bp) && !ll::is_zero(bp, bn) {
+            ll::shr(bp, bp, bn, 1);
+        }
+
+        let (_ , of) = (*ap).sub_overflow(*bp);
+        if of {
+            // (*bp - *ap) >> 1
+            ll::shr(t, &(*bp - *ap), an, 1);
```

Comment:
This will only update the first limb of `t` to `(bp - ap)/2`. You don't seem to be using anything other than the lowest limbs of anything.